File: Лабораторна №5/laba5cpp/TVector.cpp
#include <iostream>
#include <iomanip>
#include <cmath>
#include <stdlib.h>     /* srand, rand */
#include "TVector.h"

using namespace std;
//vectors
TVector::TVector(double x)
{
	this->x = x;
}
TVector::TVector(int rangeMin, int rangeMax)
{
	x = generate_number(rangeMin, rangeMax);
}
int TVector::generate_number(int rangeMin, int rangeMax)
{
	int number = rand() % (rangeMax - rangeMin + 1) + rangeMin;
	return number;
}
void TVector::display()
{
	cout << setw(4) << x << " ; ";
}


R2::R2(double x, double y) : TVector(x)
{
	this->y = y;
}
R2::R2(int rangeMin, int rangeMax) : TVector(rangeMin, rangeMax)
{
	y = generate_number(rangeMin, rangeMax);
}
void R2::display()
{
	TVector::display();
	cout << setw(4) << y << "; ";
}
bool R2::is_parallel(TVector* ptrTVector)
{
	if ((x / ptrTVector->x) == (y / ((R2*)ptrTVector)->y))
		return true;
	return false;
}
bool R2::is_perpendicular(TVector* ptrTVector)
{
	double dotProduct = x * ptrTVector->x + y * ((R2*)ptrTVector)->y;
	return !dotProduct;
}
double R2::calc_length()
{
	return sqrt(pow(x, 2) + pow(y, 2));
}


R3::R3(double x, double y, double z) : TVector(x)
{
	this->y = y;
	this->z = z;
}
R3::R3(int rangeMin, int rangeMax) : TVector(rangeMin, rangeMax)
{
	y = generate_number(rangeMin, rangeMax);
	z = generate_number(rangeMin, rangeMax);
}

void R3::display()
{
	TVector::display();
	cout << setw(4) << y << "; " << setw(4) << z << "; ";
}
bool R3::is_parallel(TVector* ptrTVector)
{
	double k = (x / ptrTVector->x);
	if (k == (y / ((R3*)ptrTVector)->y) && k == (z / ((R3*)ptrTVector)->z))
		return true;
	return false;
}
bool R3::is_perpendicular(TVector* ptrTVector)
{
	double dotProduct = x * ptrTVector->x + y * ((R3*)ptrTVector)->y + z * ((R3*)ptrTVector)->z;
	return !dotProduct;
}
double R3::calc_length()
{
	return sqrt(pow(x, 2) + pow(y, 2) + pow(z, 2));
}

File: Лабораторна №5/laba5py/TVector.py
from abc import ABC, abstractmethod
from math import sqrt

class TVector(ABC):
    
    def __init__(self, x):
        self.x = x

    def display(self):
        print(f"{self.x}; ", end = '')

    @abstractmethod
    def is_parallel(self, vector):
        pass

    @abstractmethod
    def is_perpendicular(self, vector):
        pass

    @abstractmethod
    def calc_length(self):
        pass


class R2(TVector):
    def __init__(self, x, y):
        super().__init__(x);
        self.__y = y

    def display(self):
        super().display()
        print(f"{self.__y}; ", end = '')

    def is_parallel(self, vector):
        cross_product = self.x * vector.__y - self.__y * vector.x
        if cross_product == 0:
            return True
        return False

    def is_perpendicular(self, vector):
        dot_product = self.x * vector.x + self.__y * vector.__y
        if dot_product == 0:
            return True
        return False

    def calc_length(self):        
        return sqrt(self.x ** 2 + self.__y ** 2) 


class R3(TVector):
    def __init__(self, x, y, z):
        super().__init__(x);
        self.__y = y
        self.__z = z

    def display(self):
        super().display()
        print(f"{self.__y}; {self.__z}; ", end = '')

    def is_parallel(self, vector):
        cross_product = (self.__y * vector.__z - vector.__y * self.__z) - (self.x * vector.__z - vector.x * self.__z) + (self.x * vector.__y - vector.x * self.__y)
        if cross_product == 0:
            return True
        return False

    def is_perpendicular(self, vector):
        dot_product = self.x * vector.x + self.__y * vector.__y + self.__z * vector.__z
        if dot_product == 0:
            return True
        return False

    def calc_length(self):        
        return sqrt(self.x ** 2 + self.__y ** 2 + self.__z ** 2) 
      